kernel/comps/device/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+//!     SysObj, SysPerms, SysStr,
+//! };
+//! use inherit_methods_macro::inherit_methods;
+//!
+//! #[derive(Debug)]
+//! pub struct MyDevice {
+//!     id: DeviceId,
+//!     fields: BranchNodeFields<dyn SysBranchNode, Self>,
+//! }
+//!
+//! impl Device for MyDevice {
+//!     fn device_type(&self) -> DeviceType {
+//!         DeviceType::Char
+//!     }
+//!
+//!     fn device_id(&self) -> Option<DeviceId> {
+//!         Some(self.id)
+//!     }
+//!
+//!     fn sysnode(&self) -> Arc<dyn SysBranchNode> {
+//!         self.weak_self().upgrade().unwrap()
+//!     }
+//! }
+//!
+//! inherit_sys_branch_node!(MyDevice, fields, {
+//!    fn perms(&self) -> SysPerms {
+//!        SysPerms::DEFAULT_RW_PERMS
+//!    }
+//! }
+//!
+//! #[inherit_methods(from = "self.fields")]
+//! impl MyDevice {
+//!     pub fn init_parent(&self, parent: Weak<dyn SysBranchNode>);
+//!     pub fn weak_self(&self) -> &Weak<Self>;
+//!     pub fn child(&self, name: &str) -> Option<Arc<dyn SysBranchNode>>;
+//!     pub fn add_child(&self, new_child: Arc<dyn SysBranchNode>) -> Result<()>;
+//!     pub fn remove_child(&self, child_name: &str) -> Result<Arc<dyn SysBranchNode>>;
+//! }
+//!
+//! impl MyDevice {
+//!     fn new(id: DeviceId, name: &str) -> Arc<Self> {
+//!         let name = SysStr::from(file.name().to_string());
+//!
+//!         let mut builder = SysAttrSetBuilder::new();
+//!         // Add common attributes.
+//!         builder.add(SysStr::from("dev"), SysPerms::DEFAULT_RO_ATTR_PERMS);
+//!         builder.add(SysStr::from("uevent"), SysPerms::DEFAULT_RW_ATTR_PERMS);
+//!         let attrs = builder.build().expect("Failed to build attribute set");
+//!
+//!         Arc::new_cyclic(|weak_self| MyDevice {
+//!             id,
+//!             fields: BranchNodeFields::new(name, attrs, weak_self.clone()),
+//!         })
+//!     }
+//! }
+//!
+//! // Register the device
+//! const MY_MAJOR: u32 = 1;
+//! let ida = register_device_ids(DeviceType::Char, MY_MAJOR, 0..256).unwrap();
+//! let id = ida.allocate(0).unwrap();;
+//! add_device(MyDevice::new(id, "my_device"));
+//! ```
+
+#![no_std]
+#![deny(unsafe_code)]
+#![feature(linked_list_cursors)]
+#![feature(trait_upcasting)]
+
+extern crate alloc;
+
+mod block;
+mod char;
+mod id;
+mod sysnode;
+
+use alloc::{format, sync::Arc};
+use core::fmt::Debug;
+
+use aster_systree::SysBranchNode;
+use component::{init_component, ComponentInitError};
+pub use id::{
+    register_device_ids, DeviceId, DeviceIdAllocator, DeviceType, MAJOR_BITS, MINOR_BITS,
+};
+use spin::Once;
+use sysnode::{DevNode, DevSymlink, DevSymlinks, DevicesNode};
+
+/// A trait representing a device.
+pub trait Device: Send + Sync + Debug {
+    /// Returns the type of the device.
+    fn type_(&self) -> DeviceType;
+
+    /// Returns the device ID, if the device has one.
+    fn id(&self) -> Option<DeviceId>;
+
+    /// Returns the sysfs node representing this device.
+    fn sysnode(&self) -> Arc<dyn SysBranchNode>;
```

Comment:
In Linux, every device object is a kobject. In parallel, our `Device` trait should be a sub-trait of `SysBranchNode`.